File: crates/vibra-dsp/src/modules/multiplier.rs
use super::{Module, ModuleKind, ModuleManifest, ParamDef, PortDef, SignalType, ParamUnit};

pub struct Multiplier {
    gain: f32,
}

impl Multiplier {
    pub const MANIFEST: ModuleManifest = ModuleManifest {
        id: "builtin-mult",
        name: "Multiplier",
        category: "utility",
        kind: ModuleKind::Multiplier,
        inputs: &[
            PortDef { id: "a", name: "Sound", signal_type: SignalType::Audio, accepts: &[SignalType::Audio] },
            PortDef { id: "b", name: "Mod", signal_type: SignalType::Level, accepts: &[SignalType::Level, SignalType::Modulation] },
        ],
        outputs: &[PortDef { id: "out", name: "Out", signal_type: SignalType::Audio, accepts: &[] }],
        parameters: &[
            ParamDef { id: "gain", name: "Gain", description: "Output scaling.", unit: ParamUnit::Ratio, min: 0.0, max: 2.0, default: 1.0, enum_values: &[] },
        ],
        voice_scope: super::VoiceScope::PerVoice,
        create: |_sr, _bs| Box::new(Multiplier::new()),
    };

    pub fn new() -> Self {
        Self { gain: 1.0 }
    }
}

impl Module for Multiplier {
    fn process(&mut self, inputs: &[&[f32]], outputs: &mut [&mut [f32]], frames: usize) {
        let a = &inputs[0][..frames];
        let b = if inputs.len() > 1 {
            &inputs[1][..frames]
        } else {
            &[] as &[f32]
        };
        let out = &mut outputs[0][..frames];

        for i in 0..frames {
            let mod_sig = if i < b.len() { b[i] } else { 1.0 };
            out[i] = a[i] * mod_sig * self.gain;
        }
    }

    fn set_param(&mut self, index: usize, value: f32) {
        if index == 0 {
            self.gain = value.max(0.0).min(2.0);
        }
    }

    fn set_voice(&mut self, _freq: f32, _gate: f32, _velocity: f32) {}
    fn num_inputs(&self) -> usize {
        2
    }
    fn num_outputs(&self) -> usize {
        1
    }
    fn kind(&self) -> ModuleKind {
        ModuleKind::Multiplier
    }

    fn params(&self) -> &'static [crate::param::ParamDef] {
        Self::MANIFEST.parameters
    }
}

File: frontend/src/synth/builtins/multiplier.js
export default {
  id: "builtin-mult",
  name: "Multiplier",
  plainName: "VCA / Amplitude Modulator",
  description:
    "Multiplies two signals together. Its superpower is combining audio with a control signal (like an envelope or LFO) to shape volume or create rhythmic effects.",
  whyUseIt:
    "This is how you make an envelope control loudness: patch your sound into A, and an Envelope into B. The envelope's shape will sculpt the sound's volume.",
  category: "utility",
  kind: 9,
  inputs: [
    {
      id: "a",
      name: "Sound (A)",
      description: "The main audio signal — usually an oscillator or noise.",
      signalType: "audio",
      accepts: ["audio"],
    },
    {
      id: "b",
      name: "Mod (B)",
      description: "The control signal — usually an Envelope or LFO. This gets multiplied with the sound to shape it.",
      signalType: "level",
      accepts: ["level", "modulation"],
    },
  ],
  outputs: [
    {
      id: "out",
      name: "Out",
      description: "The multiplied result. If B is an envelope, this output will be the shaped sound.",
      signalType: "audio",
      accepts: [],
    },
  ],
  parameters: [
    {
      id: "gain",
      name: "Boost",
      description: "Extra volume boost after multiplication. Use if the result is too quiet.",
      type: "float",
      min: 0,
      max: 2,
      step: 0.01,
      default: 1.0,
      paramId: 0,
    },
  ],
  create() {
    return null;
  },
  update() {},
  destroy() {},
};
